kinesis-video-pic/src/client/src/StreamEvent.c

```diff
@@ -664,18 +662,60 @@ STATUS streamTerminatedEvent(PKinesisVideoStream pKinesisVideoStream, UPLOAD_HAN
                         0,
                         0));
             }
-        }
+        } else {
 
-        // If the connection that upload handle represents is already dead. Then it will not make anymore
-        // getStreamData call so it should be removed.
-        if (!connectionStillAlive) {
-            // Remove the handle info from the queue
-            deleteStreamUploadInfo(pKinesisVideoStream, pUploadHandleInfo);
-            pUploadHandleInfo = NULL;
+            pUploadHandleInfo = getStreamUploadInfo(pKinesisVideoStream, uploadHandle);
+
+            // If the upload handle has not streamde any data, we can safely ignore this event.
+            // else the upload handle has streamed some data, set flag to trigger rollback in the next getStreamData call.
+            if (NULL != pUploadHandleInfo) {
+                if ((pUploadHandleInfo->state & UPLOAD_HANDLE_STATE_NOT_IN_USE) != UPLOAD_HANDLE_STATE_NONE) {
+                    // Need to indicate to the getStreamData to not rollback.
+                    pKinesisVideoStream->connectionState = UPLOAD_CONNECTION_STATE_NOT_IN_USE;
+                } else {
+                    // If the errored handle is the only handle, then rollback,
+                    // otherwise do not rollback because the rollback will corrupt other active handle.
+                    CHK_STATUS(stackQueueGetCount(pKinesisVideoStream->pUploadInfoQueue, &sessionCount));
+
+                    for (i = 0; i < sessionCount && !otherActiveSessionPresent; i++) {
+                        CHK_STATUS(stackQueueGetAt(pKinesisVideoStream->pUploadInfoQueue, i, &item));
+                        pUploadHandleInfoItem = (PUploadHandleInfo) item;
+                        CHK(pUploadHandleInfoItem != NULL, STATUS_INTERNAL_ERROR);
+
+                        if (pUploadHandleInfoItem != pUploadHandleInfo &&
+                            (pUploadHandleInfoItem->state & UPLOAD_HANDLE_STATE_ACTIVE) != UPLOAD_HANDLE_STATE_NONE) {
+                            otherActiveSessionPresent = TRUE;
```

Comment:
Perhaps describe here that we have the case of an old handle terminating but a new handle that's already streaming and in that case we leave it alone.

Reply:
One more thing. We should explicitly do something when we get into the state of a last ACK awaiting upload handle being dropped without durable persistence. Perhaps it's not at this layer but in C SDK layer. As we discussed, calling the error callback might be one way for now until we add a filter or have separate callback for fatal/unrecoverable errors. At least we should log.